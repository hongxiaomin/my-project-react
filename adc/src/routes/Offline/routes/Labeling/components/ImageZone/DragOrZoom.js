/* eslint no-unused-expressions: 0 */
import React from 'react'
import PropTypes from 'prop-types'
import { getScroll } from 'Utils'

export class DragOrZoom extends React.Component {
  constructor() {
    super()

    this.element = null
    this.initVal = { x: 0, y: 0 }
    this.down = false
    this.state = {
      offset: { x: 0, y: 0 },
    }
  }

  componentDidMount() {
    this.element.addEventListener('mousedown', (e) => {
      if (!e.ctrlKey) return
      e.preventDefault()
      /**
       * 1. 如果没有按ctrl不做响应
       * 2. 记录当前鼠标相对于浏览器左上角的位置, 因为使用client会受滚动条影响， 所以要加上滚动条的位置
       */
      const { scrollTop, scrollLeft } = getScroll()
      const clientX = e.clientX + scrollLeft
      const clientY = e.clientY + scrollTop
      this.initVal = {
        x: clientX / this.props.scale,
        y: clientY / this.props.scale,
        startX: this.state.offset.x,
        startY: this.state.offset.y,
      }
      this.down = true
    })
    this.element.addEventListener('mousemove', this.move)
    window.addEventListener('mouseup', (e) => {
      if (!e.ctrlKey && !this.down) return
      e.preventDefault()
      e.stopPropagation()
      this.down = false

      /**
       * 把 offset 的值传给外边
       * 更新state
       */
      this.props.onChange && this.props.onChange(this.state.offset)
    })
  }

  componentWillReceiveProps(nextProps) {
    // 这里有bug， 需要改变外边的值
    if (nextProps.scale === 1) {
      this.setState({
        offset: {
          x: 0,
          y: 0,
        },
      })
    }
  }

  move = (e) => {
    e.preventDefault()
    if (!e.ctrlKey || !this.down) return
    /**
     * 1. 用js更新样式， 不通过更新state
     */
    const { scrollTop, scrollLeft } = getScroll()
    const clientX = (e.clientX + scrollLeft) / this.props.scale
    const clientY = (e.clientY + scrollTop) / this.props.scale
    const { x, y, startX, startY } = this.initVal

    this.setState({
      offset: {
        x: startX + (clientX - x),
        y: startY + (clientY - y),
      },
    })
  }

  render() {
    const styles = {
      width: '100%',
      height: '100%',
      // position: 'relative',
      transform: `scale(${this.props.scale}) translate(${this.state.offset.x}px, ${this.state.offset.y}px)`,
      transformOrigin: '0 0',
    }
    return (<div ref={(e) => { this.element = e }} style={styles}>
      {this.props.children}
    </div>)
  }
}

DragOrZoom.propTypes = {
  scale: PropTypes.number.isRequired,
  children: PropTypes.object.isRequired,
  onChange: PropTypes.func.isRequired,
}

export default DragOrZoom
