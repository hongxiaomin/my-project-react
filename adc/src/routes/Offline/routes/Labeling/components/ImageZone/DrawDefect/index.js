import DrawDefect from './DrawDefect'
export default DrawDefect
