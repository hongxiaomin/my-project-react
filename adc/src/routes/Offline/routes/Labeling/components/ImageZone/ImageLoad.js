/* eslint jsx-a11y/alt-text: 0 */
import React from 'react'
import PropTypes from 'prop-types'
import { loadImage, getCurrentStyle } from 'Utils'

export class ImageLoad extends React.Component {

  constructor() {
    super()

    this.img = null
    this.state = {}
  }

  componentDidMount() {
    // 加载图片， 把图片信息传导外边
    this.props.imgPath && this.getImageInfo(this.props.imgPath)
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.imgPath && (nextProps.imgPath !== this.props.imgPath)) {
      this.getImageInfo(nextProps.imgPath)
    }
  }

  getImageInfo = (imgPath) => {
    loadImage(imgPath)
      .then((data) => {
        // 图片默认宽度和真实宽度的比例
        const currentWidth = parseInt(getCurrentStyle(this.img, 'width'), 10) || 0
        this.props.onChange({
          width: data.width,
          height: data.height,
          imgWidth: currentWidth,
          aspectRatio: currentWidth / data.width,
        })
      })
      .catch((e) => {
        throw new Error(`图片加载失败!\n${e}`)
      })
  }

  render() {
    if (!this.props.imgPath) {
      return null
    }

    return (<div>
      <img
        ref={(x) => { this.img = x }}
        width="100%"
        src={this.props.imgPath}
      />
      {this.props.children ? this.props.children : ''}
    </div>)
  }
}

ImageLoad.defaultProps = {
  children: undefined,
}

ImageLoad.propTypes = {
  imgPath: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  children: PropTypes.object,
}

export default ImageLoad
