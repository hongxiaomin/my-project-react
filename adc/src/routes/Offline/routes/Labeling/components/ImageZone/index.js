import ImageZone from './ImageZone'
export default ImageZone
