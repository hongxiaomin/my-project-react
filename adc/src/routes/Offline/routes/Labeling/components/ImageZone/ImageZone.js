/* eslint import/no-named-as-default: 0 */
import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { getCurrentStyle } from 'Utils'
import DragOrZoom from './DragOrZoom'
import ImageLoad from './ImageLoad'
import DrawDefect from './DrawDefect'

export class ImageZone extends React.Component {
  constructor(props) {
    super(props)

    this.keydown = false
    this.div = null
    this.imageInfo = {}
    this.state = {
      offset: { x: 0, y: 0 },
      imageInfo: { width: 0, height: 0, aspectRatio: 1 },
      boxStyle: {
        width: '100%',
        height: this.props.isUpdate ? '600px' : '300px',
      },
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.path === '') {
      this.setState({
        boxStyle: {
          width: '100%',
          height: nextProps.isUpdate ? '600px' : '300px',
        },
      })
    }
  }

  getImageInfo = ({ width, height, imgWidth, aspectRatio }) => {
    /**
     * 1. 如果height > width 得到浏览器最大高度， 计算出框的最大高度, 根据宽高比， 得到框的宽度
     * 框的宽度和图片的宽度都是根据宽高比和高度相比得来
     */

    const dHeight = document.documentElement.clientHeight || document.body.clientHeight
    const divWidth = parseInt(getCurrentStyle(this.div, 'width'), 10)

    const state = {
      imageInfo: {
        width,
        height,
        imgWidth,
        aspectRatio,
      },
    }

    let t = 60 + 40 + 50 + 40 + 15 + 10 + 10 // 除去图片展示框的其它元素的高度
    if (this.props.isUpdate) {
      t = 50 + 20
    }
    // 根据宽高比的不同， 调整展示图片的框的大小
    if (height > width) {
      const boxHeight = dHeight - t
      const boxWidth = boxHeight * (width / height)
      state.boxStyle = {
        height: boxHeight,
        width: boxWidth,
      }
      state.imageInfo.aspectRatio = boxWidth / width
    } else if (height < width) {
      let boxWidth = divWidth
      let boxHeight = divWidth * (height / width)
      if (boxHeight > (dHeight - t)) {
        boxWidth = divWidth * ((dHeight - t) / boxHeight)
        boxHeight = dHeight - t
      }
      state.boxStyle = {
        height: boxHeight,
        width: boxWidth,
      }
      state.imageInfo.aspectRatio = boxWidth / width
    } else if (height === width) {
      const boxHeight = dHeight - t
      state.boxStyle = {
        height: boxHeight,
        width: boxHeight,
      }
      state.imageInfo.aspectRatio = boxHeight / width
    }

    this.setState({
      ...state,
    })
  }

  dragOrZoomChange = (offset) => {
    this.setState({
      offset,
    })
  }

  render() {
    return (<div
      ref={(x) => { this.div = x }}
      style={{
        ...this.state.boxStyle,
        overflow: 'hidden',
        position: 'resative',
        border: '1px solid #ccc',
        margin: this.props.isUpdate ? '0 auto' : '',
      }}
    >
      <DragOrZoom
        scale={this.props.scale}
        onChange={this.dragOrZoomChange}
      >
        <ImageLoad
          imgPath={this.props.path}
          onChange={this.getImageInfo}
        >
          <DrawDefect
            scale={this.props.scale}
            offset={this.state.offset}
            imageInfo={this.state.imageInfo}
            div={this.div}
          />
        </ImageLoad>
      </DragOrZoom>
    </div>)
  }
}

ImageZone.propTypes = {
  scale: PropTypes.number.isRequired,
  path: PropTypes.string.isRequired,
  isUpdate: PropTypes.bool,
}

ImageZone.defaultProps = {
  isUpdate: false,
}

export default connect(({ labelingTool }) => ({
  scale: labelingTool.controlPanel.zoom,
  path: labelingTool.selectedImageInfo.path,
}))(ImageZone)
