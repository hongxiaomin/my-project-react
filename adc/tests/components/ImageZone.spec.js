import React from 'react'
import { shallow, mount } from 'enzyme'
import { ImageZone } from 'routes/Offline/routes/Labeling/components/ImageZone/ImageZone'

describe('(Component) ImageZone', () => {
  const scale = 0.5
  const path = ''
  const isUpdate = true
  const wrapper = shallow(
    <ImageZone
      scale={scale}
      path={path}
      isUpdate={isUpdate}
    />,
  )
  it('element DragOrZoom', () => {
    expect(wrapper.find('DragOrZoom').length).toBe(1)
  })
  it('element ImageLoad', () => {
    expect(wrapper.find('ImageLoad').length).toBe(1)
  })
  wrapper.find('DragOrZoom').simulate('change', { x: 1, y: 1 })
  expect(wrapper.state('offset')).toEqual({ x: 1, y: 1 })
  wrapper.find('ImageLoad').simulate('change', { width: 30, height: 50, imageWidth: 20, aspectRatio: 0.2 })
  expect(wrapper.state('offset')).toEqual({ x: 1, y: 1 })
  wrapper.find('ImageLoad').simulate('change', { width: 60, height: 50, imageWidth: 20, aspectRatio: 0.2 })
  expect(wrapper.state('offset')).toEqual({ x: 1, y: 1 })
  wrapper.find('ImageLoad').simulate('change', { width: 50, height: 50, imageWidth: 20, aspectRatio: 0.2 })
  expect(wrapper.state('offset')).toEqual({ x: 1, y: 1 })
  it('render normally', () => {
    expect(wrapper).toMatchSnapshot()
  })
})

describe('(Component) mount', () => {
  const scale = 0.5
  const path = ''
  const isUpdate = true
  const wrapper = mount(
    <ImageZone
      scale={scale}
      path={path}
      isUpdate={isUpdate}
    />,
  )
  it('render normally', () => {
    expect(wrapper).toMatchSnapshot()
  })
  it('element DragOrZoom', () => {
    expect(wrapper.find('DragOrZoom').length).toBe(1)
  })
  it('element ImageLoad', () => {
    expect(wrapper.find('ImageLoad').length).toBe(1)
  })
})
describe('(Component) mount', () => {
  const scale = 0.5
  const path = ''
  const isUpdate = false
  const wrapper = mount(
    <ImageZone
      scale={scale}
      path={path}
      isUpdate={isUpdate}
    />,
  )
  it('render isUpdate = false', () => {
    expect(wrapper).toMatchSnapshot()
  })
  it('element DragOrZoom', () => {
    expect(wrapper.find('DragOrZoom').length).toBe(1)
  })
  it('element ImageLoad', () => {
    expect(wrapper.find('ImageLoad').length).toBe(1)
  })
})