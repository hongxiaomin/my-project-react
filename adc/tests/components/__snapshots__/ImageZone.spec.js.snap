// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) ImageZone render normally 1`] = `
<div
  style={
    Object {
      "border": "1px solid #ccc",
      "height": -70,
      "margin": "0 auto",
      "overflow": "hidden",
      "position": "resative",
      "width": -70,
    }
  }
>
  <DragOrZoom
    onChange={[Function]}
    scale={0.5}
  >
    <ImageLoad
      imgPath=""
      onChange={[Function]}
    >
      <Connect(DrawDefect)
        div={null}
        imageInfo={
          Object {
            "aspectRatio": -1.4,
            "height": 50,
            "imgWidth": undefined,
            "width": 50,
          }
        }
        offset={
          Object {
            "x": 1,
            "y": 1,
          }
        }
        scale={0.5}
      />
    </ImageLoad>
  </DragOrZoom>
</div>
`;

exports[`(Component) mount render isUpdate = false 1`] = `
<ImageZone
  isUpdate={false}
  path=""
  scale={0.5}
>
  <div
    style={
      Object {
        "border": "1px solid #ccc",
        "height": "300px",
        "margin": "",
        "overflow": "hidden",
        "position": "resative",
        "width": "100%",
      }
    }
  >
    <DragOrZoom
      onChange={[Function]}
      scale={0.5}
    >
      <div
        style={
          Object {
            "height": "100%",
            "transform": "scale(0.5) translate(0px, 0px)",
            "transformOrigin": "0 0",
            "width": "100%",
          }
        }
      >
        <ImageLoad
          imgPath=""
          onChange={[Function]}
        />
      </div>
    </DragOrZoom>
  </div>
</ImageZone>
`;

exports[`(Component) mount render normally 1`] = `
<ImageZone
  isUpdate={true}
  path=""
  scale={0.5}
>
  <div
    style={
      Object {
        "border": "1px solid #ccc",
        "height": "600px",
        "margin": "0 auto",
        "overflow": "hidden",
        "position": "resative",
        "width": "100%",
      }
    }
  >
    <DragOrZoom
      onChange={[Function]}
      scale={0.5}
    >
      <div
        style={
          Object {
            "height": "100%",
            "transform": "scale(0.5) translate(0px, 0px)",
            "transformOrigin": "0 0",
            "width": "100%",
          }
        }
      >
        <ImageLoad
          imgPath=""
          onChange={[Function]}
        />
      </div>
    </DragOrZoom>
  </div>
</ImageZone>
`;
